### PR TITLE
[TACHYON-694][TACHYON-695] Fine grained locking in tiered storage

### PR DESCRIPTION
**Pull request by apc999**
https://tachyon.atlassian.net/browse/TACHYON-694
https://tachyon.atlassian.net/browse/TACHYON-695

Current tiered storage (`TieredBlockStore`) has read/write lock "mEvictionLock" as a global lock (within a single store) to provide thread safety. While simple, it comprises performance. In this PR, we avoid using global lock but try use per-Block lock as much as possible as a performance optimization.

### CODE REVIEW

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 1]
```diff
@@ -74,34 +84,30 @@
   private final Evictor mEvictor;
   private final List<BlockStoreEventListener> mBlockStoreEventListeners =
       new ArrayList<BlockStoreEventListener>();
-  /** A set of pinned inodes fetched from the master */
+  /**
+   * A set of pinned inodes fetched from the master
+   */
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Could you use the old style for this?

File: servers/src/main/java/tachyon/worker/block/BlockLockManager.java [hunk 2]
```diff
@@ -60,24 +58,22 @@
...
+  public BlockLockManager() {
     for (int i = 0; i < NUM_LOCKS; i ++) {
       mLockArray[i] = new ClientRWLock();
     }
   }
 
   /**
-   * Locks a block if it exists, throws NotFoundException otherwise.
+   * Locks a block. Note that, lock striping is used so even this block does not exist, a lock id
+   * is still returned.
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/BlockLockManager.java]: Does this make the logic simpler?

Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockLockManager.java]: Right, this change is to simplify the logic: otherwise we need to meta-lock `mMetaManager` in this method to check the existence of the given block.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 1 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Done thanks

**Comment by apc999** [comment]
@shimingfei , could you take a look at this PR? thanks!

**Comment by shimingfei** [comment]
@apc999 
Sorry for late reply, I will take a look today.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 3]
```diff
@@ -126,73 +135,93 @@ public void unlockBlock(long userId, long blockId) throws NotFoundException {
...
-      mEvictionLock.writeLock().unlock();
+    for (int i = 0; i < MAX_RETRIES; i ++) {
+      TempBlockMeta tempBlockMeta =
+          createBlockMetaInternal(userId, blockId, location, initialBlockSize);
+      if (tempBlockMeta != null) {
+        return tempBlockMeta;
+      }
+      // Failed to allocate a temp block, so trigger Evictor to make some space Note that, even
+      // {@link freeSpaceInternal} succeeds, it does not ensure the next allocation successful,
+      // because these two operations are not atomic.
```
Comment (shimingfei) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: To make sure there will be enough space after eviction, the evictor should evict exactly the space size requested, not "requested space - rest space" in some directory.

**Comment by shimingfei** [comment]
@apc999 
it seems that the deadlock issues in 0.7 branch still exists in current code, right?

File: common/src/main/java/tachyon/util/io/BufferUtils.java [hunk 4]
```diff
@@ -148,14 +149,19 @@ public static ByteBuffer getIncreasingIntBuffer(int len) {
   }
 
   /**
-   * * Writes buffer to the given file path
+   * Writes buffer to the given file path. Overwrites this file if exists.
```
Comment (gpang) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: I think technically, this method deletes the file first, if it already exists, not overwriting it.

File: servers/src/main/java/tachyon/worker/block/BlockMetadataManager.java [hunk 5]
```diff
@@ -335,6 +335,18 @@ public synchronized boolean hasTempBlockMeta(long blockId) {
     return false;
   }
 
+  public BlockMeta moveBlockMeta(BlockMeta blockMeta, TempBlockMeta tempBlockMeta)
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/BlockMetadataManager.java]: javadoc this public method?

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 6]
```diff
@@ -58,14 +59,21 @@
  * Allocator to decide where to put a new block, an Evictor to decide where to evict a stale block,
  * a BlockMetadataManager to maintain the status of the tiered storage, and a LockManager to
  * coordinate read/write on the same block.
- * <p>
+ * <p/>
  * This class is thread-safe.
+ * <p/>
+ * Locking hierarchy
+ * <p/>
+ * A read/write lock to ensure eviction is atomic w.r.t. other operations. An eviction may trigger a
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Is this comment block true anymore? It sounds like the old, big eviction lock comment.

Here, the new locking scheme should be clearly described.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 7]
```diff
@@ -126,73 +135,93 @@ public void unlockBlock(long userId, long blockId) throws NotFoundException {
   @Override
   public BlockWriter getBlockWriter(long userId, long blockId) throws NotFoundException,
       IOException {
-    TempBlockMeta tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
-    return new LocalFileBlockWriter(tempBlockMeta);
+    // NOTE, a temp block is supposed to be visible for one writer
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: for "only" one writer? Is this the reason that the block lock does not need to be acquired? That should be explicit in the comment.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 8]
```diff
@@ -126,73 +135,93 @@ public void unlockBlock(long userId, long blockId) throws NotFoundException {
...
-      return createBlockMetaNoLock(userId, blockId, location, initialBlockSize);
-    } finally {
-      mEvictionLock.writeLock().unlock();
+    for (int i = 0; i < MAX_RETRIES; i ++) {
+      TempBlockMeta tempBlockMeta =
+          createBlockMetaInternal(userId, blockId, location, initialBlockSize);
+      if (tempBlockMeta != null) {
+        return tempBlockMeta;
+      }
+      // Failed to allocate a temp block, so trigger Evictor to make some space Note that, even
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Add a period after "space". Also, add "if" after "even".

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 9]
```diff
@@ -342,104 +368,208 @@ public void registerBlockStoreEventListener(BlockStoreEventListener listener) {
...
    * @param userId the ID of user
    * @param blockId the ID of block
+   * @return the {@link tachyon.worker.block.meta.TempBlockMeta} of blockId
    * @throws NotFoundException if blockId can not be found in temporary blocks
    * @throws AlreadyExistsException if blockId already exists in committed blocks
    * @throws InvalidStateException if blockId is not owned by userId
-   * @return the {@link tachyon.worker.block.meta.TempBlockMeta} of blockId
    */
-  private TempBlockMeta validateUserTempBlock(long userId, long blockId) throws NotFoundException,
+  private TempBlockMeta checkTempBlockExists(long userId, long blockId) throws NotFoundException,
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: This seems more like a "getTempBlock" or "getUserTempBlock" or something. What do you think?

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 10]
```diff
@@ -342,104 +368,208 @@ public void registerBlockStoreEventListener(BlockStoreEventListener listener) {
...
+      BlockStoreLocation loc;
+      String srcPath;
+      String dstPath;
+      mMetadataLock.writeLock().lock();
+      try {
+        TempBlockMeta tempBlockMeta = checkTempBlockExists(userId, blockId);
+        srcPath = tempBlockMeta.getPath();
+        dstPath = tempBlockMeta.getCommitPath();
+        loc = tempBlockMeta.getBlockLocation();
+        mMetaManager.commitTempBlockMeta(tempBlockMeta);
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: There is a gap in time before the commit of the temp block, and the move to the committed path, so something could happen between the two operations. Will this be fine?

I want to say "yes", since the block lock is (usually) acquired before the metadata lock? What about the scenarios where the block lock is NOT acquired at all (but the metadata lock is)?

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 11]
```diff
@@ -342,104 +368,208 @@ public void registerBlockStoreEventListener(BlockStoreEventListener listener) {
...
+   * @param blockId block Id
+   * @param location location to create the block
+   * @param initialBlockSize initial block size in bytes
+   * @return a temp block created if successful, or null if allocation failed (instead of throwing
+   *         OutOfSpaceException because allocation failure could be an expected case)
+   * @throws AlreadyExistsException if there is a block already having the same block id
+   */
+  private TempBlockMeta createBlockMetaInternal(long userId, long blockId,
+      BlockStoreLocation location, long initialBlockSize) throws AlreadyExistsException {
+    mMetadataLock.writeLock().lock();
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: This doesn't need the block lock first?

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 12]
```diff
@@ -342,104 +368,208 @@ public void registerBlockStoreEventListener(BlockStoreEventListener listener) {
...
+  /**
+   * Increases the temp block size, and this will succeed only if there is enough available space in
+   * this temp block's parent dir. Returns a pair of boolean and BlockStoreLocation where the
+   * boolean indicates if the operation succeeds and the BlockStoreLocation denotes where to free
+   * more space if it fails.
+   */
+  private Pair<Boolean, BlockStoreLocation> requestSpaceInternal(long blockId, long additionalBytes)
+      throws NotFoundException, OutOfSpaceException, IOException, AlreadyExistsException,
+      InvalidStateException {
+    mMetadataLock.writeLock().lock();
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: This doesn't need the block lock first?

File: servers/src/main/java/tachyon/worker/block/meta/BlockMeta.java [hunk 13]
```diff
@@ -30,7 +30,8 @@ public BlockMeta(long blockId, long blockSize, StorageDir dir) {
 
   public BlockMeta(TempBlockMeta tempBlock) {
     super(tempBlock.getBlockId(), tempBlock.getParentDir());
-    mBlockSize = new File(tempBlock.getCommitPath()).length();
+    // NOTE: TempBlockMeta must be committed before the actual data block file is moved.
+    mBlockSize = new File(tempBlock.getPath()).length();
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/meta/BlockMeta.java]: Why does this use getPath instead of getCommitPath?

**Comment by gpang** [comment]
Thanks for the PR @apc999 ! I left some comments and questions in the review.

File: common/src/main/java/tachyon/util/io/BufferUtils.java (hunk 4 above)
Comment (apc999) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: Fixed.

File: servers/src/main/java/tachyon/worker/block/BlockMetadataManager.java (hunk 5 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockMetadataManager.java]: done

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 6 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Good call. More comments added.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 7 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Done.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 8 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Done

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 9 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Fixed by having this method returning void.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 11 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: A comment added

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 12 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Same reason as `createBlockMetaInternal`, a comment added.

File: servers/src/main/java/tachyon/worker/block/meta/BlockMeta.java (hunk 13 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/meta/BlockMeta.java]: Reverted back, as we will do IO before the final commit.

**Comment by apc999** [comment]
@shimingfei do you mean the deadlock caused by synchronized clients? I believe we don't have this issue as now (1) the block lock is acquired outside the meta lock, (2) meta lock is never held across methods and (3) in this new version, we lock only with best effort. I could be very wrong, so please let me know if you find any cases

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 14]
```diff
@@ -126,73 +144,94 @@ public void unlockBlock(long userId, long blockId) throws NotFoundException {
...
+      // NOTE: Successful {@link freeSpaceInternal} here does not ensure the next try of allocation
+      // also successful, because these two operations are not atomic.
+      if (i < MAX_RETRIES - 1) {
+        freeSpaceInternal(userId, initialBlockSize, location);
+      }
     }
+    // TODO: we are probably seeing a rare transient failure, maybe define and throw some other
+    // types of exception to indicate this case.
+    throw new OutOfSpaceException("Failed to create blockMeta: blockId " + blockId + " "
+        + "failed to allocate " + initialBlockSize + " bytes after " + MAX_RETRIES + " retries");
```
Comment (carsonwang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: If it means RETRY, we should have 3 + 1 tries by default? Otherwise MAX_TRIES will be more accurate.

**Comment by shimingfei** [comment]
@apc999 
The JIRA about this issue is https://tachyon.atlassian.net/browse/TACHYON-709
eg: two blocks(block1 and block2) share the same lock(mapping to the same index in BlockLockManager), and there are two threads(thread1 and thread2) on client side, block1 is locked(readlock) by  thread1. Then thread2 requests space from worker, and the worker needs to evict some block to get enough space,  block2 needs to be evicted, worker needs to get write lock of block2, but it will be blocked, because there is read lock on the lock.
there are two ways to solve the problem:
1. PR1197 against to TACHYON-709 (Fix deadlock because different blocks may share the same lock on worker side)
OR
2. use try lock. just like PR1231 against to TACHYON-731(use try lock in worker)

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 15]
```diff
@@ -200,16 +239,15 @@ public void abortBlock(long userId, long blockId) throws AlreadyExistsException,
...
-    } finally {
-      mEvictionLock.writeLock().unlock();
+    for (int i = 0; i < MAX_RETRIES; i ++) {
+      Pair<Boolean, BlockStoreLocation> requestResult =
+          requestSpaceInternal(blockId, additionalBytes);
+      if (requestResult.getFirst()) {
+        return;
+      }
+      if (i < MAX_RETRIES - 1) {
+        freeSpaceInternal(userId, additionalBytes, requestResult.getSecond());
```
Comment (carsonwang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Do we want to retry if `requestSpaceInternal` or `freeSpaceInternal` throws an exception? If so,   we need catch the exceptions except for the last try.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 16]
```diff
@@ -342,104 +372,224 @@ public void registerBlockStoreEventListener(BlockStoreEventListener listener) {
...
       try {
-        removeBlockNoLock(userId, blockId);
-        synchronized (mBlockStoreEventListeners) {
-          for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
-            listener.onRemoveBlockByWorker(userId, blockId);
-          }
+        removeBlockInternal(userId, blockId);
+      } catch (NotFoundException nfe) {
+        LOG.info("Failed to evict blockId " + blockId + ", it could be already deleted");
+        return;
```
Comment (carsonwang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Why don't we continue if it has already been deleted?

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 17]
```diff
@@ -454,26 +604,26 @@ private void freeSpaceInternal(long userId, long availableBytes, BlockStoreLocat
...
-        long lockId = mLockManager.lockBlock(userId, blockId, BlockLockType.WRITE);
+        MoveBlockResult moveResult;
         try {
-          moveBlockNoLock(blockId, newLocation);
-          synchronized (mBlockStoreEventListeners) {
-            for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
-              listener.onMoveBlockByWorker(userId, blockId, oldLocation, newLocation);
-            }
+          // TODO: this should also specify the src location
+          moveResult = moveBlockInternal(userId, blockId, newLocation);
```
Comment (carsonwang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Do we need check the moveResult here?

**Comment by carsonwang** [comment]
@apc999 , the patch looks cool. I believe it helps improving the performance and also avoid some dead lock situations we observed before like TACHYON-731. I just left some comments and questions.

**Comment by apc999** [comment]
@shimingfei Regarding the bug you mentioned above (https://tachyon.atlassian.net/browse/TACHYON-709), yes this PR still have this problem. My tentative plan is to let @calvinjia will fix it in v0.7 branch, and we will soon merge that back to master. If you think there is a better / easier way, please let me know.

**Comment by shimingfei** [comment]
@apc999 That sounds OK, if the PR will be merged back to master.

File: common/src/main/java/tachyon/util/io/BufferUtils.java [hunk 18]
```diff
@@ -148,14 +149,19 @@ public static ByteBuffer getIncreasingIntBuffer(int len) {
   }
 
   /**
-   * * Writes buffer to the given file path
+   * Writes buffer to the given file path. If this file already exists, deletes it first.
    *
    * @param path file path to write the data
    * @param buffer raw data
-   * @throws java.io.IOException
+   * @throws java.io.IOException if an I/O error occurred or failed to delete existing file
```
Comment (calvinjia) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: Maybe just `@throws IOException`?

File: common/src/main/java/tachyon/util/io/BufferUtils.java [hunk 19]
```diff
@@ -148,14 +149,19 @@ public static ByteBuffer getIncreasingIntBuffer(int len) {
...
    */
   public static void writeBufferToFile(String path, byte[] buffer) throws IOException {
-    FileOutputStream os = new FileOutputStream(path);
+    File file = new File(path);
+    if (file.exists()) {
+      FileUtils.delete(file);
+    }
+    FileUtils.createFile(file);
+    FileOutputStream os = new FileOutputStream(file);
```
Comment (calvinjia) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: I think you can just do `new FileOutputStream(file, false)` to overwrite the existing file.

File: servers/src/main/java/tachyon/worker/block/BlockDataManager.java [hunk 20]
```diff
@@ -220,7 +220,7 @@ public void commitBlock(long userId, long blockId) throws AlreadyExistsException
     } catch (TException te) {
       throw new IOException("Failed to commit block to master.", te);
     } finally {
-      mBlockStore.unlockBlock(userId, blockId);
+      mBlockStore.unlockBlock(lockId);
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/BlockDataManager.java]: Why is the userId omitted here?

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 21]
```diff
@@ -126,73 +144,94 @@ public void unlockBlock(long userId, long blockId) throws NotFoundException {
   @Override
   public BlockWriter getBlockWriter(long userId, long blockId) throws NotFoundException,
       IOException {
-    TempBlockMeta tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
-    return new LocalFileBlockWriter(tempBlockMeta);
+    // NOTE: a temp block is supposed to be visible for its own writer, unnecessary to acquire
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Maybe specify `supposed to only be visible`

Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Also, are we planning to handle multiple writers to the same block in this PR? (basically making the temp blocks keyed off user id and block id instead of just block id).

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 14 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Good point. I restructured the code to take this into account.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 22]
```diff
@@ -126,73 +144,94 @@ public void unlockBlock(long userId, long blockId) throws NotFoundException {
...
+          createBlockMetaInternal(userId, blockId, location, initialBlockSize, true);
+      if (tempBlockMeta != null) {
+        return tempBlockMeta;
+      }
+      // Failed to allocate a temp block, so trigger Evictor to make some space.
+      // NOTE: Successful {@link freeSpaceInternal} here does not ensure the next try of allocation
+      // also successful, because these two operations are not atomic.
+      if (i < MAX_RETRIES - 1) {
+        freeSpaceInternal(userId, initialBlockSize, location);
+      }
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: I think you can move freeSpaceInternal before the createBlockMetaInternal since the eviction should no-op if there is already space? Just makes the code easier to reason about I think.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 15 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: This is really a good call. I cleaned a little bit the exceptions thrown by helper functions in this file. For this particular method, I prefer not retrying as now those exceptions are basically very bad. 

Meanwhile, in the future, we may want to differentiate exceptions for transient or permanent reasons and making retry based on this. This is already aimed in JIRA TACHYON-521

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 16 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Fixed. Thanks.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 17 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Done.

File: common/src/main/java/tachyon/util/io/BufferUtils.java (hunk 19 above)
Comment (apc999) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: The problem is if parent dir of the target file is not constructed yet, the constructor will fail

File: common/src/main/java/tachyon/util/io/BufferUtils.java (hunk 18 above)
Comment (apc999) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: Done.

File: servers/src/main/java/tachyon/worker/block/BlockDataManager.java (hunk 20 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockDataManager.java]: Currently two methods are equivalent in function. But `mBlockStore.unlockBlock(userId, blockId);` is added tentatively and deprecated and `mBlockStore.unlockBlock(lockId);` is the preferred way?

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 21 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: No plan to handle this case in this PR. I have created a JIRA (TACHYON-747) for this case.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 22 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: I have restructured the code to better make retry of `MAX_RETRIES` times.
Let me know if you think it is still not as easy as you expected

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 23]
```diff
@@ -275,63 +306,67 @@ public void accessBlock(long userId, long blockId) throws NotFoundException {
...
       for (StorageDir dir : tier.getStorageDirs()) {
         File userFolder = new File(PathUtils.concatPath(dir.getDirPath(), userId));
-        if (userFolder.exists() && !userFolder.delete()) {
-          // This error means we could not delete the directory but should not affect the
+        try {
+          if (userFolder.exists()) {
+            FileUtils.delete(userFolder);
+          }
+        } catch (IOException ioe) {
+          // This error means we could not delete the directory but should not affect t©©he
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: clean up the last "the" on the line.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 24]
```diff
@@ -539,4 +824,38 @@ public void updatePinnedInodes(Set<Integer> inodes) {
...
+
+    MoveBlockResult(boolean done, long blockSize, BlockStoreLocation srcLocation,
+        BlockStoreLocation dstLocation) {
+      mDone = done;
+      mBlockSize = blockSize;
+      mSrcLocation = srcLocation;
+      mDstLocation = dstLocation;
+    }
+
+    boolean done() {
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Is this basically success or not? I think "success()" (and mSuccess) would be a better name for it.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 25]
```diff
@@ -493,38 +697,119 @@ private BlockMetadataManagerView getUpdatedView() {
...
-    if (mMetaManager.hasTempBlockMeta(blockId)) {
-      throw new InvalidStateException("Failed to move block " + blockId + ": block is uncommited");
+  /**
+   * Moves a block to new location only if allocator finds available space in newLocation. This
+   * method will not trigger any eviction. Returns MoveBlockResult.
+   *
+   * @param userId user Id
+   * @param blockId block Id
+   * @param newLocation new location to move this block
+   * @return move result
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Maybe "the resulting information about the move operation"

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 26]
```diff
@@ -539,4 +824,38 @@ public void updatePinnedInodes(Set<Integer> inodes) {
...
+      mBlockSize = blockSize;
+      mSrcLocation = srcLocation;
+      mDstLocation = dstLocation;
+    }
+
+    boolean done() {
+      return mDone;
+    }
+
+    long blockSize() {
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Why does the result of a move need the block size?

**Comment by gpang** [comment]
I left a few comments/questions.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 24 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Done

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 25 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Done

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 26 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: If this move fails, we need the block size to tell `freeSpaceInternal` how much space to make

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 23 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: done

File: common/src/main/java/tachyon/util/io/BufferUtils.java (hunk 19 above)
Comment (calvinjia) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: Wasn't that also the behavior before? Or are we trying to improve this util to create the file regardless of that?

File: servers/src/main/java/tachyon/worker/block/BlockDataManager.java (hunk 20 above)
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/BlockDataManager.java]: Got it, thanks.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 22 above)
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Its still a bit weird, are you concerned about the performance impact of pessimistically running `freeSpaceInternal`?

**Comment by calvinjia** [comment]
@apc999 Looks good overall, just two minor comments left.

**Comment by gpang** [comment]
LGTM

File: common/src/main/java/tachyon/util/io/BufferUtils.java (hunk 19 above)
Comment (apc999) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: You are right that the current behavior is same as before. This change is due to a change I reverted already in this PR. Let me revert this change also to keep this PR smaller and cleaner.

File: servers/src/test/java/tachyon/worker/block/meta/BlockMetaTest.java [hunk 27]
```diff
@@ -15,6 +15,7 @@
 
 package tachyon.worker.block.meta;
 
+import java.io.File;
```
Comment (uronce-cc) [inline servers/src/test/java/tachyon/worker/block/meta/BlockMetaTest.java]: unused import?

Comment (apc999) [inline servers/src/test/java/tachyon/worker/block/meta/BlockMetaTest.java]: Thanks. Fixed.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java (hunk 22 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: Right. Because a read lock will still be required even under a pessimistic run of `freeSpaceInternal`. Therefore I try to introduce less lock contention here.

**Comment by apc999** [comment]
@uronce-cc @calvinjia  comments fixed.

**Comment by uronce-cc** [comment]
+1 LGTM!

**Comment by calvinjia** [comment]
+1 LGTM
